alloc.h

```diff
@@ -19,52 +22,124 @@ void* kalloc(size_t sz);
 void kfree(void* ptr);
 
 struct page {
-    // YOUR CODE HERE
+    list_links link;
+    int order;
+    uintptr_t address;
+    bool free = 1;
 };
 
+list <page, &page :: link> orders[22]; //free list
+page pageArray[NPAGES + 1];
 
 uintptr_t getBuddy(uintptr_t pa, int o)
 {
-    // YOUR CODE HERE
+    return pa ^ (1 << o);
 }
 
+//freelist[o].front() first thing
+
 // init_kalloc
 //    Initialize stuff needed by `kalloc`. Should be called
 //    in testpset1.cpp before using 'kalloc' and 'bfree'. 
 void init_kalloc() {
-    // YOUR CODE HERE
+    pageArray[0].order = 21;
+    for (int index = 0; index <= 512; index++){
+        pageArray[index].address = index * 4096;
+    }
+    pageArray[0].free = 1;
+    orders[21].push_back(&pageArray[0]);
+    cout << "init alloc done!! awesome" << endl;
 }
 
-
 // kalloc(sz)
 //    Allocate and return a pointer to at least `sz` contiguous bytes of
 //    memory. Returns `nullptr` if `sz == 0` or on failure.
 //
 //    The caller should initialize the returned memory before using it.
 //    The handout allocator sets returned memory to 0xCC 
+
 void* kalloc(size_t sz) 
 {
-    // REWRITE 
-    if (sz > PAGESIZE) return nullptr;
+    //special case 
+    if (sz == 0) return nullptr;
 
     void* ptr = nullptr;
+    cout << "ALLOCATING" << endl;
 
-    if (next_free_pa < (1UL << 21))
-    {
-        ptr = pa2kptr(next_free_pa);
-        next_free_pa += PAGESIZE;
+    int order = msb(sz - 1);
+    if (order < 12){
+        order = 12; 
+    }
+    int currentOrder = order;
+
+    while (currentOrder <= 21){
+        //empty means FULL!!
+        //special case for 21
+        if (currentOrder == 21 && !orders[21].empty()){
+            cout << "SPECIAL CASE 21" << endl;
+            page* page = orders[currentOrder].pop_front();
+            int index = page->address / 4096;
+            pageArray[index].free = 0;
+            return page;
+        }
+
+        //searching for nearest unoccupied order
+        while (orders[currentOrder].empty()){
+            currentOrder++;
+            if (currentOrder > 21){
+                cout << "no free blocks" << endl;
+                return ptr;
+            }
+        }
+        while (currentOrder >= order){
```

Comment:
this loop should handle the order == 21 case as well - no need for the special case